quetz/main.py

```diff
@@ -1630,65 +1628,74 @@ def serve_path(
     if is_package_request and channel.mirror_channel_url:
-        # if we exclude the package from syncing, redirect to original URL
         channel_proxylist = json.loads(channel.channel_metadata).get('proxylist', [])
         if channel_proxylist and package_name and package_name in channel_proxylist:
             return RedirectResponse(f"{channel.mirror_channel_url}/{path}")
 
+    fsize = fmtime = fetag = None
+
     if channel.mirror_channel_url and channel.mirror_mode == "proxy":
         repository = RemoteRepository(channel.mirror_channel_url, session)
-        if not pkgstore.file_exists(channel.name, path):
+        if is_repodata_request:
+            # Invalidate repodata.json and current_repodata.json after channel.ttl seconds
+            try:
+                fsize, fmtime, fetag = pkgstore.get_filemetadata(channel.name, path)
+                cache_miss = time.time() - fmtime >= channel.ttl
+            except FileNotFoundError:
+                cache_miss = True
+        else:
+            cache_miss = not pkgstore.file_exists(channel.name, path)
+        if cache_miss:
             download_remote_file(repository, pkgstore, channel.name, path)
-        elif path.endswith(".json"):
-            # repodata.json and current_repodata.json are cached locally
-            # for channel.ttl seconds
-            _, fmtime, _ = pkgstore.get_filemetadata(channel.name, path)
-            if time.time() - fmtime >= channel.ttl:
-                download_remote_file(repository, pkgstore, channel.name, path)
-
-    if (
-        is_package_request or pkgstore.kind == "LocalStore"
-    ) and pkgstore.support_redirect:
-        return RedirectResponse(pkgstore.url(channel.name, path))
-
-    def iter_chunks(fid):
-        while True:
-            data = fid.read(chunk_size)
-            if not data:
-                break
-            yield data
-
-    if path == "" or path.endswith("/"):
-        path += "index.html"
-    package_content_iter = None
-
-    headers = {}
-    if accept_encoding and 'gzip' in accept_encoding and path.endswith('.json'):
-        # return gzipped response
-        try:
-            package_content_iter = iter_chunks(
-                pkgstore.serve_path(channel.name, path + '.gz')
-            )
-            path += '.gz'
-            headers['Content-Encoding'] = 'gzip'
-            headers['Content-Type'] = 'application/json'
-        except FileNotFoundError:
-            pass
+            fsize = fmtime = fetag = None
 
-    while not package_content_iter:
-        try:
-            package_content_iter = iter_chunks(pkgstore.serve_path(channel.name, path))
+    gzip_exists = (
+        is_repodata_request
+        and accept_encoding
+        and 'gzip' in accept_encoding
+        and pkgstore.file_exists(channel.name, path + ".gz")
+    )
 
+    # Redirect response
+    if (is_package_request or is_repodata_request) and pkgstore.support_redirect:
+        return RedirectResponse(
```

Comment:
did you confirm that this works? Since it redirects to a path with `.gz` attached and I am not sure mamba / conda will handle that correctly.